sharding/hashValidatorShuffler.go

```diff
@@ -380,26 +383,54 @@ func removeLeavingNodesFromValidatorMaps(
 	copy(stillRemainingInLeaving, leaving)
 
 	if !waitingFixEnabled {
-		waiting, stillRemainingInLeaving = removeNodesFromMap(waiting, stillRemainingInLeaving, numToRemove)
-		eligible, stillRemainingInLeaving = removeNodesFromMap(eligible, stillRemainingInLeaving, numToRemove)
-	} else {
-		maxNumToRemoveFromWaiting := make(map[uint32]int)
-		for i := range numToRemove {
-			maxNumToRemoveFromWaiting[i] = len(eligible[i]) + len(waiting[i]) - minNodesPerShard
+		newWaiting, stillRemainingInLeaving := removeNodesFromMap(waiting, stillRemainingInLeaving, numToRemove)
+		newEligible, stillRemainingInLeaving := removeNodesFromMap(eligible, stillRemainingInLeaving, numToRemove)
+		return newEligible, newWaiting, stillRemainingInLeaving
+	}
+
+	return removeLeavingNodes(eligible, waiting, numToRemove, stillRemainingInLeaving, minNodesMeta, minNodesPerShard)
+}
+
+func removeLeavingNodes(
+	eligible map[uint32][]Validator,
+	waiting map[uint32][]Validator,
+	numToRemove map[uint32]int,
+	stillRemainingInLeaving []Validator,
+	minNodesMeta int,
+	minNodesPerShard int,
+) (map[uint32][]Validator, map[uint32][]Validator, []Validator) {
+	maxNumToRemoveFromWaiting := make(map[uint32]int)
+	for shardId := range numToRemove {
+		minimumNumberOfNodes := minNodesPerShard
```

Comment:
Duplicated code with L418. 
Maybe we can figure out a way to refactor these components as to remove the number of passed parameters and the duplicated code.

Reply:
fixed